libs/libc/time/lib_strptime.c

```diff
@@ -0,0 +1,540 @@
+              {
+                if (tm->tm_hour > 12) /* i.e., 13:00 AM ?! */
+                  return (NULL);
+                else if (tm->tm_hour == 12)
+                  tm->tm_hour = 0;
+
+                bp += len;
+                break;
+              }
+
+            /* PM? */
+
+            len = strlen(_ctloc(am_pm[1]));
+            if (strncasecmp(_ctloc(am_pm[1]), (const char *)bp, len) == 0)
+              {
+                if (tm->tm_hour > 12) /* i.e., 13:00 PM ?! */
+                  return (NULL);
+                else if (tm->tm_hour < 12)
+                  tm->tm_hour += 12;
+
+                bp += len;
+                break;
+              }
+
+            /* Nothing matched. */
+
+            return (NULL);
+
+          case 'S': /* The seconds. */
+            _LEGAL_ALT(_ALT_O);
+            if (!(_conv_num(&bp, &tm->tm_sec, 0, 61)))
+              return (NULL);
+            break;
+
+          case 'U': /* The week of year, beginning on sunday. */
+          case 'W': /* The week of year, beginning on monday. */
+            _LEGAL_ALT(_ALT_O);
+            /* XXX This is bogus, as we can not assume any valid
+            * information present in the tm structure at this
+            * point to calculate a real value, so just check the
+            * range for now.
+            */
+
+            if (!(_conv_num(&bp, &i, 0, 53)))
+              return (NULL);
+            break;
+
+          case 'w': /* The day of week, beginning on sunday. */
+            _LEGAL_ALT(_ALT_O);
+            if (!(_conv_num(&bp, &tm->tm_wday, 0, 6)))
+              return (NULL);
+            break;
+
+          case 'Y': /* The year. */
+            _LEGAL_ALT(_ALT_E);
+            if (!(_conv_num(&bp, &i, 0, 9999)))
+              return (NULL);
+
+            cr->relyear = -1;
+            tm->tm_year = i - TM_YEAR_BASE;
+            break;
+
+          case 'y': /* The year within the century (2 digits). */
+            _LEGAL_ALT(_ALT_E | _ALT_O);
+            if (!(_conv_num(&bp, &cr->relyear, 0, 99)))
+              return (NULL);
+            break;
+
+            /* Miscellaneous conversions. */
+
+          case 'n': /* Any kind of white-space. */
+          case 't':
+            _LEGAL_ALT(0);
+            while (isspace(*bp))
+              bp++;
+            break;
+
+          default: /* Unknown/unsupported conversion. */
+            return (NULL);
+        }
+    }
+
+  /* We need to evaluate the two digit year spec (%y)
+   * last as we can get a century spec (%C) at any time.
+   */
+
+  if (cr->relyear != -1)
+  {
+    if (cr->century == TM_YEAR_BASE)
+    {
+      if (cr->relyear <= 68)
+        tm->tm_year = cr->relyear + 2000 - TM_YEAR_BASE;
+      else
+        tm->tm_year = cr->relyear + 1900 - TM_YEAR_BASE;
+    }
+    else
+    {
+      tm->tm_year = cr->relyear + cr->century - TM_YEAR_BASE;
+    }
+  }
+
+  return (unsigned char *)bp;
```

Comment:
why `const` is discarded?

Reply:
> > why `const` is discarded?
> 
> since _strptime and strptime return non-const pointer

Yeah, that is obvious. My question is if `bp` points to RO memory (flash) and the result is casted then caller may `p = strptime(...)` -> `p[0] = '1';` that will lead to crash. I hope that will not happen. I understand that thre is no `_r` variant, but probably the global holder should be used in this case and not dropping of `const`.